File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/BlockChangePacket.java
package gg.mineral.server.network.packet.play.clientbound;

import gg.mineral.api.network.packet.Packet;
import io.netty.buffer.ByteBuf;

public final record BlockChangePacket(int x, short y, int z, int blockId, short blockMetadata)
        implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        os.writeInt(x);
        os.writeByte(y);
        os.writeInt(z);
        writeVarInt(os, blockId);
        os.writeByte(blockMetadata);
    }

    @Override
    public byte getId() {
        return 0x23;
    }
}

File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/BlockChangePacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf

@JvmRecord
data class BlockChangePacket(val x: Int, val y: Short, val z: Int, val blockId: Int, val blockMetadata: Short) :
    Packet.OUTGOING {
    override fun serialize(os: ByteBuf) {
        os.writeInt(x)
        os.writeByte(y.toInt())
        os.writeInt(z)
        writeVarInt(os, blockId)
        os.writeByte(blockMetadata.toInt())
    }

    override fun getId(): Byte {
        return 0x23
    }
}
